File: merge-tonic-lib/tonic/hydration/hydration_config.py
"""Hydration YAML/JSON config (parity with hydrationConfig.ts)."""

from __future__ import annotations

import json
import re
from dataclasses import dataclass
from pathlib import Path
from typing import Any, Literal

QuestionRefinementMode = Literal["off", "improver", "subquestions"]
QuestionRefinementContext = Literal["minimal", "progressive"]


@dataclass
class ResolvedHydrationConfig:
    question_mode: QuestionRefinementMode
    refinement_context: QuestionRefinementContext
    strict_llm: bool
    llm_model: str
    llm_base_url: str
    openai_api_key_env: str
    llm_json_object: bool = True


def load_hydration_config_file(file_path: str | None) -> dict[str, Any]:
    if not (file_path or "").strip():
        return {}
    p = Path(file_path)
    raw = p.read_text(encoding="utf-8")
    if str(p).lower().endswith(".json"):
        return json.loads(raw)
    out: dict[str, Any] = {}
    for line in raw.splitlines():
        m = re.match(r"^(\w+):\s*(.*)$", line.strip())
        if not m:
            continue
        k, v = m.group(1), m.group(2).strip().strip("\"'")
        if k == "question_mode" and v in ("off", "improver", "subquestions"):
            out["question_mode"] = v
        elif k == "question_refinement_context" and v in ("minimal", "progressive"):
            out["question_refinement_context"] = v
        elif k == "strict_llm":
            out["strict_llm"] = v in ("true", "1", "yes")
        elif k == "llm_model":
            out["llm_model"] = v
        elif k == "llm_base_url":
            out["llm_base_url"] = v
        elif k == "openai_api_key_env":
            out["openai_api_key_env"] = v
        elif k == "llm_json_object":
            out["llm_json_object"] = v in ("true", "1", "yes")
    return out


def resolve_hydration_config(
    config_path: str | None,
    env: dict[str, str],
    *,
    question_mode_cli: QuestionRefinementMode | None = None,
    strict_llm: bool | None = None,
    llm_model: str | None = None,
    llm_base_url: str | None = None,
    openai_api_key_env: str | None = None,
) -> ResolvedHydrationConfig:
    base = ResolvedHydrationConfig(
        question_mode="off",
        refinement_context="minimal",
        strict_llm=False,
        llm_model="gpt-4o-mini",
        llm_base_url="https://api.openai.com/v1",
        openai_api_key_env="OPENAI_API_KEY",
        llm_json_object=True,
    )
    file_cfg = load_hydration_config_file(config_path)
    qm: QuestionRefinementMode = base.question_mode
    if file_cfg.get("question_mode") in ("off", "improver", "subquestions"):
        qm = file_cfg["question_mode"]
    env_mode = (env.get("TONIC_QUESTION_MODE") or "").strip()
    if env_mode in ("off", "improver", "subquestions"):
        qm = env_mode  # type: ignore[assignment]
    if question_mode_cli is not None:
        qm = question_mode_cli

    rc: QuestionRefinementContext = (
        "progressive" if file_cfg.get("question_refinement_context") == "progressive" else base.refinement_context
    )

    sl = base.strict_llm
    if "strict_llm" in file_cfg:
        sl = bool(file_cfg["strict_llm"])
    if strict_llm is not None:
        sl = strict_llm

    jo = base.llm_json_object
    if "llm_json_object" in file_cfg:
        jo = bool(file_cfg["llm_json_object"])
    env_jo = (env.get("TONIC_LLM_JSON_OBJECT") or "").strip().lower()
    if env_jo in ("0", "false", "no", "off"):
        jo = False
    if env_jo in ("1", "true", "yes", "on"):
        jo = True

    return ResolvedHydrationConfig(
        question_mode=qm,
        refinement_context=rc,
        strict_llm=sl,
        llm_model=(llm_model or file_cfg.get("llm_model") or base.llm_model).strip() or base.llm_model,
        llm_base_url=(llm_base_url or file_cfg.get("llm_base_url") or base.llm_base_url).strip()
        or base.llm_base_url,
        openai_api_key_env=(
            openai_api_key_env or file_cfg.get("openai_api_key_env") or base.openai_api_key_env
        ).strip()
        or base.openai_api_key_env,
        llm_json_object=jo,
    )

File: packages/tonic-core/src/hydration/hydrationConfig.ts
import * as fs from "node:fs";

export type QuestionRefinementMode = "off" | "improver" | "subquestions";
export type QuestionRefinementContext = "minimal" | "progressive";

export type HydrationConfigFile = {
  question_mode?: QuestionRefinementMode;
  question_refinement_context?: QuestionRefinementContext;
  strict_llm?: boolean;
  llm_model?: string;
  llm_base_url?: string;
  openai_api_key_env?: string;
  /** When false, omit json_object response_format for chat completions. */
  llm_json_object?: boolean;
};

export type ResolvedHydrationConfig = {
  questionMode: QuestionRefinementMode;
  refinementContext: QuestionRefinementContext;
  strictLlm: boolean;
  llmModel: string;
  llmBaseUrl: string;
  openaiApiKeyEnv: string;
  llmJsonObject: boolean;
};

const defaults: ResolvedHydrationConfig = {
  questionMode: "off",
  refinementContext: "minimal",
  strictLlm: false,
  llmModel: "gpt-4o-mini",
  llmBaseUrl: "https://api.openai.com/v1",
  openaiApiKeyEnv: "OPENAI_API_KEY",
  llmJsonObject: true,
};

export function loadHydrationConfigFile(filePath: string | undefined): Partial<HydrationConfigFile> {
  if (!filePath?.trim()) {
    return {};
  }
  const raw = fs.readFileSync(filePath, "utf8");
  const ext = filePath.toLowerCase();
  if (ext.endsWith(".json")) {
    return JSON.parse(raw) as HydrationConfigFile;
  }
  // Minimal YAML subset: key: value lines (no nested objects required for v1)
  const out: HydrationConfigFile = {};
  for (const line of raw.split(/\r?\n/)) {
    const m = /^(\w+):\s*(.*)$/.exec(line.trim());
    if (!m) {
      continue;
    }
    const k = m[1]!;
    const v = m[2]!.replace(/^["']|["']$/g, "").trim();
    if (k === "question_mode" && (v === "off" || v === "improver" || v === "subquestions")) {
      out.question_mode = v;
    }
    if (k === "question_refinement_context" && (v === "minimal" || v === "progressive")) {
      out.question_refinement_context = v;
    }
    if (k === "strict_llm") {
      out.strict_llm = v === "true" || v === "1" || v === "yes";
    }
    if (k === "llm_model") {
      out.llm_model = v;
    }
    if (k === "llm_base_url") {
      out.llm_base_url = v;
    }
    if (k === "openai_api_key_env") {
      out.openai_api_key_env = v;
    }
    if (k === "llm_json_object") {
      const b = parseYamlBool(v);
      if (b !== undefined) {
        out.llm_json_object = b;
      }
    }
  }
  return out;
}

function parseYamlBool(v: string): boolean | undefined {
  if (v === "true" || v === "1" || v === "yes") {
    return true;
  }
  if (v === "false" || v === "0" || v === "no") {
    return false;
  }
  return undefined;
}

export function mergeHydrationConfig(
  base: ResolvedHydrationConfig,
  file: Partial<HydrationConfigFile>,
  env: NodeJS.ProcessEnv,
  cli: Partial<ResolvedHydrationConfig>,
): ResolvedHydrationConfig {
  const envMode = env.TONIC_QUESTION_MODE?.trim();
  let questionMode = cli.questionMode ?? base.questionMode;
  if (file.question_mode) {
    questionMode = file.question_mode;
  }
  if (envMode === "off" || envMode === "improver" || envMode === "subquestions") {
    questionMode = envMode;
  }
  if (cli.questionMode) {
    questionMode = cli.questionMode;
  }

  const envJson = env.TONIC_LLM_JSON_OBJECT?.trim().toLowerCase();
  let llmJsonObject = cli.llmJsonObject ?? base.llmJsonObject;
  if (typeof file.llm_json_object === "boolean") {
    llmJsonObject = file.llm_json_object;
  }
  if (envJson === "0" || envJson === "false" || envJson === "no" || envJson === "off") {
    llmJsonObject = false;
  }
  if (envJson === "1" || envJson === "true" || envJson === "yes" || envJson === "on") {
    llmJsonObject = true;
  }
  if (cli.llmJsonObject !== undefined) {
    llmJsonObject = cli.llmJsonObject;
  }

  return {
    questionMode,
    refinementContext:
      cli.refinementContext ??
      (file.question_refinement_context === "progressive" ? "progressive" : base.refinementContext),
    strictLlm: cli.strictLlm ?? file.strict_llm ?? base.strictLlm,
    llmModel: cli.llmModel ?? file.llm_model ?? base.llmModel,
    llmBaseUrl: cli.llmBaseUrl ?? file.llm_base_url ?? base.llmBaseUrl,
    openaiApiKeyEnv: cli.openaiApiKeyEnv ?? file.openai_api_key_env ?? base.openaiApiKeyEnv,
    llmJsonObject,
  };
}

export function resolveHydrationConfig(
  configPath: string | undefined,
  env: NodeJS.ProcessEnv,
  cliOverrides: Partial<ResolvedHydrationConfig>,
): ResolvedHydrationConfig {
  const file = loadHydrationConfigFile(configPath);
  return mergeHydrationConfig(defaults, file, env, cliOverrides);
}
